instrumentation-genai/opentelemetry-instrumentation-google-genai/tests/generate_content/test_e2e.py

```diff
@@ -0,0 +1,507 @@
+# limitations under the License.
+
+"""High level end-to-end test of the generate content instrumentation.
+
+The primary purpose of this test is to verify that the instrumentation
+package does not break the underlying GenAI SDK that it instruments.
+
+This test suite also has some minimal validation of the instrumentation
+outputs; however, validating the instrumentation output (other than
+verifying that instrumentation does not break the GenAI SDK) is a
+secondary goal of this test. Detailed testing of the instrumentation
+output is the purview of the other tests in this directory."""
+
+import asyncio
+import gzip
+import json
+import os
+import subprocess
+import sys
+
+import google.auth
+import google.auth.credentials
+import google.genai
+import pytest
+import yaml
+from vcr.record_mode import RecordMode
+
+from opentelemetry.instrumentation.google_genai import (
+    GoogleGenAiSdkInstrumentor,
+)
+
+from ..common.auth import FakeCredentials
+from ..common.otel_mocker import OTelMocker
+
+_FAKE_PROJECT = "test-project"
+_FAKE_LOCATION = "test-location"
+_FAKE_API_KEY = "test-api-key"
+_DEFAULT_REAL_LOCATION = "us-central1"
+
+
+def _get_project_from_env():
+    return (
+        os.getenv("GCLOUD_PROJECT") or os.getenv("GOOGLE_CLOUD_PROJECT") or ""
+    )
+
+
+def _get_project_from_gcloud_cli():
+    try:
+        gcloud_call_result = subprocess.run(
+            "gcloud config get project",
+            shell=True,
+            capture_output=True,
+            check=True,
+        )
+    except subprocess.CalledProcessError:
+        return None
+    gcloud_output = gcloud_call_result.stdout.decode()
+    return gcloud_output.strip()
+
+
+def _get_project_from_credentials():
+    _, from_creds = google.auth.default()
+    return from_creds
+
+
+def _get_real_project():
+    from_env = _get_project_from_env()
+    if from_env:
+        return from_env
+    from_cli = _get_project_from_gcloud_cli()
+    if from_cli:
+        return from_cli
+    return _get_project_from_credentials()
+
+
+def _get_location_from_env():
+    return (
+        os.getenv("GCLOUD_LOCATION")
+        or os.getenv("GOOGLE_CLOUD_LOCATION")
+        or ""
+    )
+
+
+def _get_real_location():
+    return _get_location_from_env() or _DEFAULT_REAL_LOCATION
+
+
+def _get_vertex_api_key_from_env():
+    return os.getenv("GOOGLE_API_KEY")
+
+
+def _get_gemini_api_key_from_env():
+    return os.getenv("GEMINI_API_KEY")
+
+
+def _should_redact_header(header_key):
+    if header_key.startswith("x-goog"):
+        return True
+    if header_key.startswith("sec-goog"):
+        return True
+    if header_key in ["server", "server-timing"]:
+        return True
+    return False
+
+
+def _redact_headers(headers):
+    to_redact = []
+    for header_key in headers:
+        if _should_redact_header(header_key.lower()):
+            to_redact.append(header_key)
+    for header_key in to_redact:
+        headers[header_key] = "<REDACTED>"
```

Comment:
```suggestion
    for header_key in headers:
        if _should_redact_header(header_key.lower()):
            headers[header_key] = "<REDACTED>"
```

Reply:
Done. I had previously done it with two loops, because of deleting the keys rather than setting the value to `<REDACTED>` which had created issues with modifying the iterator during iteration.